validator/src/admin_rpc_service.rs

```diff
@@ -48,6 +55,7 @@ pub struct AdminRpcRequestMetadata {
     pub tower_storage: Arc<dyn TowerStorage>,
     pub staked_nodes_overrides: Arc<RwLock<HashMap<Pubkey, u64>>>,
     pub post_init: Arc<RwLock<Option<AdminRpcRequestMetadataPostInit>>>,
+    pub plugin_manager: Arc<RwLock<GeyserPluginManager>>,
```

Comment:
admin service doesn't need a reference to the plugin manager.  instead, plugin manager and admin services should share (opposing ends of) a pair of [channels](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html).  plugin manager now gets a request receiver and a reply sender on init, spools up a thread and waits for requests to service. admin service gets the request sender and reply receiver, then forwards requests and replies via the admin socket.

this will allow us to encapsulate all of the plugin logic in the plugin manager where it belongs, rather than scatter it all over the place. we also avoid exposing plugin manager internals that don't need to be (`pub` fields are evil)

Reply:
I urge you to consider the benefits of the current setup over encapsulation. 

1) Save an entire thread spawn + having to periodically check for updates. As is, this doesn't affect the performance of the plugin_manager when the admin rpc service is unused. The admin rpc service is currently given 3 worker threads*** so this would be putting those to better use.
2) better reload latency vs the additional channel send/recv pings.
3) the `pub libs` aren't actually used for anything after startup and should actually probably just be dropped after calling `lib.get(b"_create_plugin")?` when loading a plugin. The only thing a lib can do is call `lib.get(b"_create_plugin")?` which should put the whole vtable in memory. As I'm writing this, i'm realizing that if we choose to keep them for whatever reason we may want to consider rebuilding the old plugin by calling `current_lib.get(b"_create_plugin")?` and then calling `on_load` versus just calling `on_load`.
4) It is possible to get a bit more encapsulation without additional thread spawns by writing the reload logic as a method of the plugin manager which is called in the admin rpc service.

*** 1) What

Reply:
Agreed with @t-nelson : no need to build direct dependency of admin_rpc_service on plugin manager. It only needs a send handle to the channel.

Reply:
You are asking the admin RPC service to manage plugins; instead of adding a dependency on the plugin manager***, you would rather have a runtime performance hit with thread spawn + channels? 

***the admin rpc service is defined in `solana-validator` which is already downstream from `solana-core` which which is already downstream from `solana-geyser-plugin-manager`. This isn't really adding a new dependency. We will always have
```
solana-validator v1.16.0
├── solana-core v1.16.0
│   ├── solana-geyser-plugin-manager v1.16.0
```
unless you plan to rip the manager out of `solana-core`.

Reply:
If you **really really** want another thread to do this, I will concede upon confirmation. But I think it's a bad idea.

Reply:
> If you **really really** want another thread to do this, I will concede upon confirmation. But I think it's a bad idea.

Yes. That is a widely used pattern in solana validator to break reduce dependencies across modules. The administrator RPC does not need to know who is actually handling that event. Suppose in one day I have wide user use case and may need to change to another handler -- we do not need to change the RPC code at all.  As long as the thread waiting for the event with good timeout like recv_timeout I do not think any performance penalties -- the OS should be able do good scheduling to use the CPU for such light activity. In addition, we do not want rpc handler to handle this synchronously any way as the operation can takes sometime.

Reply:
I mentioned this approach in the comment https://github.com/solana-labs/solana/pull/30352#issuecomment-1435062751

Reply:
If I define some enum (see below), in which crate do I put it to minimize dependencies? If I put it in the manager plugin crate, then the validator is still dependent on the manager plugin crate. Do I put it in some existing common or utils crate? If so, can you point me to it?
```rust
pub enum PluginManagerRequest {
    ReloadPlugin {
        name: String,
        libpath: String,
        config_file: String,
    },
    ListPlugins,
}
```

Reply:
Just chiming in to agree with Trent and Lijun. I do still hope to disentangle the admin-rpc-service from the solana-validator crate one day, and while there are a couple wrinkles to doing so already, I’d rather not add another one.

Reply:
Yes, we need to define the event type regardless and I would just define it pub enum PluginRequest and define in geyser-plugin-manager crate and put it in geyser_plugin_service.rs. This is still better than expose the fat class GeyserPluginManager out. You can see the existing code out side of geyser-plugin-manager do not need to know its implementation.

Reply:
> and define in geyser-plugin-manager crate

We might want to create a new `solana-admin-rpc-api` crate to hold types like this. Wdyt, @lijunwangs ?

Reply:
If introducing a new crate, I think it would be best to position it to hold other common types. Maybe call it solana-common